source/Mlos.Agent/MainAgent.cs

```diff
@@ -171,17 +173,67 @@ public static void RegisterSettingsAssembly(uint assemblyIndex)
             {
                 MlosProxyInternal.RegisteredSettingsAssemblyConfig assemblyConfig = assemblySharedConfig.Config;
 
-                // Try to load assembly from the agent folder.
+                // Start looking for places to find the assembly.
                 //
-                string applicationFolderPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
-                string assemblyFilePath = Path.Combine(applicationFolderPath, assemblyConfig.AssemblyFileName.Value);
+                List<string> assemblyDirs = new List<string>();
 
-                if (!File.Exists(assemblyFilePath))
+                // 1. Try to load assembly from the agent folder.
+                //
+                assemblyDirs.Add(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location));
+
+                // 2. Try to load assembly from the full path listed in the config.
+                // Note: This doesn't currently work for Linux.
+                // See Also: Mlos.Core/MlosContext.cpp
+                //
+                if (assemblyConfig.ApplicationFilePath.Value != null)
+                {
+                    assemblyDirs.Add(Path.GetDirectoryName(assemblyConfig.ApplicationFilePath.Value));
+                }
+
+                // 3. The current working directory.
+                //
+                assemblyDirs.Add(".");
+
+                // 4. The search path specified in an environment variable.
+                // This is akin to an LD_LIBRARY_PATH but specifically for MLOS Settings Registry DLLs.
+                //
+                string settingsRegistryLibraryPath = System.Environment.GetEnvironmentVariable("MLOS_SETTINGS_REGISTRY_PATH");
+                if (!string.IsNullOrEmpty(settingsRegistryLibraryPath))
+                {
+                    if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                    {
+                        assemblyDirs.AddRange(settingsRegistryLibraryPath.Split(';'));
+                    }
+                    else if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+                    {
+                        assemblyDirs.AddRange(settingsRegistryLibraryPath.Split(':'));
+                    }
+                    else
+                    {
+                        throw new NotImplementedException(
+                            string.Format("Unhandled OS: '{0}'", System.Runtime.InteropServices.RuntimeInformation.OSDescription ?? "Unknown"));
+                    }
+                }

```

Comment:
Could turn this first section into a yield function to avoid computing all of them if there's a match early on in the set.